bundle/deploy/terraform/state_pull.go

```diff
@@ -24,71 +28,95 @@ func (l *statePull) Name() string {
 	return "terraform:state-pull"
 }
 
-func (l *statePull) remoteState(ctx context.Context, f filer.Filer) (*bytes.Buffer, error) {
-	// Download state file from filer to local cache directory.
-	remote, err := f.Read(ctx, TerraformStateFileName)
+func (l *statePull) remoteState(ctx context.Context, b *bundle.Bundle) (*tfState, []byte, error) {
+	f, err := l.filerFactory(b)
 	if err != nil {
-		// On first deploy this state file doesn't yet exist.
-		if errors.Is(err, fs.ErrNotExist) {
-			return nil, nil
-		}
-		return nil, err
+		return nil, nil, err
 	}
 
-	defer remote.Close()
-
-	var buf bytes.Buffer
-	_, err = io.Copy(&buf, remote)
+	r, err := f.Read(ctx, TerraformStateFileName)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
+	defer r.Close()
 
-	return &buf, nil
-}
+	content, err := io.ReadAll(r)
+	if err != nil {
+		return nil, nil, err
+	}
 
-func (l *statePull) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
-	f, err := l.filerFactory(b)
+	state := &tfState{}
+	err = json.Unmarshal(content, state)
 	if err != nil {
-		return diag.FromErr(err)
+		return nil, nil, err
 	}
 
+	return state, content, nil
+}
+
+func (l *statePull) localState(ctx context.Context, b *bundle.Bundle) (*tfState, error) {
 	dir, err := Dir(ctx, b)
 	if err != nil {
-		return diag.FromErr(err)
+		return nil, err
 	}
 
-	// Download state file from filer to local cache directory.
-	log.Infof(ctx, "Opening remote state file")
-	remote, err := l.remoteState(ctx, f)
+	content, err := os.ReadFile(filepath.Join(dir, TerraformStateFileName))
 	if err != nil {
-		log.Infof(ctx, "Unable to open remote state file: %s", err)
-		return diag.FromErr(err)
+		return nil, err
 	}
-	if remote == nil {
-		log.Infof(ctx, "Remote state file does not exist")
-		return nil
+
+	state := &tfState{}
+	err = json.Unmarshal(content, state)
+	if err != nil {
+		return nil, err
 	}
 
-	// Expect the state file to live under dir.
-	local, err := os.OpenFile(filepath.Join(dir, TerraformStateFileName), os.O_CREATE|os.O_RDWR, 0600)
+	return state, nil
+}
+
+func (l *statePull) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	dir, err := Dir(ctx, b)
 	if err != nil {
 		return diag.FromErr(err)
 	}
-	defer local.Close()
 
-	if !IsLocalStateStale(local, bytes.NewReader(remote.Bytes())) {
-		log.Infof(ctx, "Local state is the same or newer, ignoring remote state")
+	localStatePath := filepath.Join(dir, TerraformStateFileName)
+
+	// Case: remote state file does not exist. In this case we should not use the
+	// local state file because we cannot guarantee it corresponds to the same deployment,
+	// that is the same root_path and workspace host.
+	remoteState, remoteContent, err := l.remoteState(ctx, b)
+	if errors.Is(err, fs.ErrNotExist) {
+		log.Infof(ctx, "Remote state file does not exist. Invalidating local terraform state.")
+		os.Remove(localStatePath)
```

Comment:
There are two cases where this breaks:
* You're intentionally changing the root path and want the state to be carried forward.
* You did a deployment and the state upload failed. This will remove the local state and orphan resources.

What are examples where this is the right behavior?

Reply:
This is the right behaviour if we do not (or can not) carry forward the deployment. Example:
1. Changing identity. I deploy as a service principal, and then as myself to get a different deployment.
2. Changing host URL. If I do so, state cannot be carried forward.

I validate that the first case does indeed fail today. Did not verify if terraform automatically handles case two though. Checking.

Reply:
> You did a deployment and the state upload failed. This will remove the local state and orphan resources.

This is only a problem for the first deployment of the DAB. For subsequent deployments, we'll correctly retain the local cache based on the sequence number.

Reply:
I think the right way to think about this is:
1. Remote is the source of truth, including the absence of remote state being an indication of no active deployment.
2. Maintaining the local cache for recovering from "failure to upload state" is a best-effort problem. Only when we are sure that remote is outdated (same lineage, lower seq number) should we defer to the local state.

Reply:
I agree with these points, but think we should not deal with them here.

1. If you change the identity or workspace, we should invalidate the entire cache, not just this file.
2. If you change the root path, you should still be able to deploy. This is possible today and allows someone to relocate a bundle (though you need to clean up the old location yourself). This change makes this impossible.

If we keep this change focused on only checking the lineage, these other points can be addressed in a dedicated change.

Reply:
Done. Removed the deletion of local state on when remote state is missing. We'll followup with the broader changes needed.